std/machines/small_field/arith.asm

```diff
@@ -96,13 +48,32 @@ machine Arith(byte: Byte, byte2: Byte2) with
         0
     };
 
-    let limbs_to_int: expr[] -> int = query |limbs| array::sum(array::map_enumerated(limbs, |i, limb| int(eval(limb)) << (i * 8)));
+    let limbs_to_int: fe[] -> int = |limbs| array::sum(array::map_enumerated(limbs, |i, limb| int(limb) << (i * 8)));
+    let int_to_limbs: int -> fe[] = |x| array::new(4, |i| fe(select_limb(x, i)));
+
+    // We need to provide hints for the quotient and remainder, because they are not unique under our current constraints.
+    // They are unique given additional main machine constraints, but it's still good to provide hints for the solver.
+    query |i| std::prover::compute_from_multi_if(
+        is_division = 1,
+        y1 + x2,
+        i,
+        y3 + x1,
+        |values| {
+            let y3_value = limbs_to_int([values[0], values[1], values[2], values[3]]);
+            let x1_value = limbs_to_int([values[4], values[5], values[6], values[7]]);
+            if x1_value == 0 {
+                // Quotient is unconstrained, use zero for y1
+                // and set remainder x2 = y3.
+                [0, 0, 0, 0] + int_to_limbs(y3_value)
+            } else {
+                let quotient = y3_value / x1_value;
+                let remainder = y3_value % x1_value;
+                int_to_limbs(quotient) + int_to_limbs(remainder)
+            }
+        }
+    );
```

Comment:
I have reviewed the mathematics behind all the computation and it makes sense (it is even clearer than before). What I don't quite understand is the use of this query function. I understand that the results are assigned in y3 + x1, but when is it “called” for that to happen?

Reply:
In the runtime solver, (all) query functions are executed when there is no more progress, I think. It works because `eval` silently fails if the value is not there yet. If all `eval` calls succeed, then the computed value is assigned to the target columns.